crypto/hpke/hpke.c

```diff
@@ -0,0 +1,1537 @@
+/**< "HPKE" - "suite_id" label for section 5.1 */
+static const char OSSL_HPKE_SEC51LABEL[] = "\x48\x50\x4b\x45";
+/**< "psk_id_hash" - in key_schedule_context */
+static const char OSSL_HPKE_PSKIDHASH_LABEL[] = "\x70\x73\x6b\x5f\x69\x64\x5f\x68\x61\x73\x68";
+/**<  "info_hash" - in key_schedule_context */
+static const char OSSL_HPKE_INFOHASH_LABEL[] = "\x69\x6e\x66\x6f\x5f\x68\x61\x73\x68";
+/**<  "base_nonce" - base nonce calc label */
+static const char OSSL_HPKE_NONCE_LABEL[] = "\x62\x61\x73\x65\x5f\x6e\x6f\x6e\x63\x65";
+/**<  "exp" - internal exporter secret generation label */
+static const char OSSL_HPKE_EXP_LABEL[] = "\x65\x78\x70";
+/**<  "sec" - external label for exporting secret */
+static const char OSSL_HPKE_EXP_SEC_LABEL[] = "\x73\x65\x63";
+/**<  "key" - label for use when generating key from shared secret */
+static const char OSSL_HPKE_KEY_LABEL[] = "\x6b\x65\x79";
+/**<  "psk_hash" - for hashing PSK */
+static const char OSSL_HPKE_PSK_HASH_LABEL[] = "\x70\x73\x6b\x5f\x68\x61\x73\x68";
+/**<  "secret" - for generating shared secret */
+static const char OSSL_HPKE_SECRET_LABEL[] = "\x73\x65\x63\x72\x65\x74";
+
+/* "strength" input to RAND_bytes_ex */
+#define OSSL_HPKE_RSTRENGTH 10
+
+/**
+ * @brief sender or receiver context
+ */
+struct ossl_hpke_ctx_st
+{
+    OSSL_LIB_CTX *libctx; /**< library context */
+    char *propq; /**< properties */
+    int mode; /**< HPKE mode */
+    OSSL_HPKE_SUITE suite; /**< suite */
+    uint64_t seq; /**< sequence number */
+    unsigned char *shared_secret; /* KEM output, zz */
+    size_t shared_secretlen;
+    unsigned char *key; /* final aead key */
+    size_t keylen;
+    unsigned char *nonce; /* aead base nonce */
+    size_t noncelen;
+    unsigned char *exportersec; /**< exporter secret */
+    size_t exporterseclen;
+    char *pskid; /**< PSK stuff */
+    unsigned char *psk;
+    size_t psklen;
+    EVP_PKEY *authpriv; /**< sender's authentication private key */
+    unsigned char *authpub; /**< auth public key */
+    size_t authpublen;
+    unsigned char *ikme; /* IKM if using deterministic key gen */
+    size_t ikmelen;
+};
+
+/**
+ * @brief check if KEM uses NIST curve or not
+ * @param kem_id is the externally supplied kem_id
+ * @return 1 for NIST, 0 for good-but-non-NIST, other otherwise
+ */
+static int hpke_kem_id_nist_curve(uint16_t kem_id)
+{
+    const OSSL_HPKE_KEM_INFO *kem_info = NULL;
+
+    kem_info = ossl_HPKE_KEM_INFO_find_id(kem_id);
+    return kem_info != NULL && kem_info->groupname != NULL;
+}
+
+/**
+ * @brief wrapper to import NIST curve public key as easily as x25519/x448
+ * @param libctx is the context to use
+ * @param propq is a properties string
+ * @param gname is the curve groupname
+ * @param buf is the binary buffer with the (uncompressed) public value
+ * @param buflen is the length of the private key buffer
+ * @return a working EVP_PKEY * or NULL
+ */
+static EVP_PKEY *EVP_PKEY_new_raw_nist_public_key(OSSL_LIB_CTX *libctx,
```

Comment:
ossl_pkey_new_raw_ec_public_key

Reply:
see above wrt NIST, resolution here will follow that

Reply:
question though: I chose  ``EVP_PKEY_new_raw_nist_public_key()`` as the CFRG-curve equivalent is ``EVP_PKEY_new_raw_public_key_ex()`` so not sure why you'd prefer ossl_pkey_* instead?